### ISSUE TITLE
Additional smoke tests for release 2.1

### ISSUE THREAD

**Issue by ewallace** [riboviz/riboviz#355]
Many things have changed in preparation for release 2.1 and I think we need to do some ~integration~ smoke tests to ensure that the entire pipeline still works without nasty surprises.

Given our paper's focus on running across the tree of life, I suggest that sufficient integration tests would include:
* one yeast dataset
* one bacterial dataset
* one other organism dataset

This could be combined with [Align configurations with RiboViz release 2.1 configurations, example-datasets#76](/riboviz/example-datasets/issues/76).

**Comment by mikej888** [comment]
All pending 2.1 tasks need to be completed and merged into `develop` before this can begin. At the last developer meeting this was scheduled to be done by Wed 16 June. So that leaves Thu 17 June for people to do these tests and for any issues to be resolved for a release on Fr 18 June.

**Comment by mikej888** [comment]
What I back in the past erroneously termed regression tests are in fact integration tests - see #290. If a run on a full size dataset raises an then this means that our automated tests are inadequate and need extended.

**Comment by ewallace** [comment]
Apologies, I used the wrong terminology in referring to these as integration tests and have renamed the issue to "smoke tests". It is important to check on full-size datasets exactly to check for any unexpected bugs and, if necessary, write new automated tests.

Discussed with @FlicAnderson today and she will address.

**Comment by FlicAnderson** [comment]
I'll be running these on Eddie this afternoon or tomorrow morning after #367 and then #366 are merged into develop and I get the nod from @mikej888. 

I'll try and set up some of the files required tonight to save time. 

After these run (hopefully successfully), we're good to proceed to Release 2.1! :crossed_fingers:

**Comment by FlicAnderson** [comment]
Merged @acope3's PR #368 into develop after running tests as well.

Now currently running updated riboviz `develop` branch on Eddie on L-Sc_2014 (fungi) dataset (config in example-datasets branch `update-configs-76`).

I'm also currently setting up configuration files & job G-Hv_2020 (archaea) and W-Ec_2015 (bacteria) to get those jobs ready to run & submit.

**Comment by FlicAnderson** [comment]
I'm getting some issues on Eddie while trying to run more than one riboviz job submission at once - specifically trying to get G-Hv_2020 (archaea) and W-Ec_2015 (bacteria) running.  L-Sc_2014 is currently running and has no issues so far.  

I had tested these G & W datasets via validate_only and such in an interactive session for each of these failing configs, so I'm fairly sure they *should* run now, and in the same interactive session I tested that they successfully got to the first process when I gave the 'full run' command, but I didn't necessarily wait about too long in an interactive session because there was lots to be getting on with.  

This is the error from the job .e file after the job submission for the full dataset 'completed' (unsuccessfully) for G-Hv_2020, but it's the same story for W-Ec_2015: 
```
Error occurred during initialization of VM
Could not allocate metaspace: 1073741824 bytes
NOTE: Nextflow is trying to use the Java VM defined by the following environment variables:
 JAVA_CMD: /exports/csce/eddie/biology/groups/wallace_rna/anaconda/envs/riboviz/bin/java
 NXF_OPTS:

```

It doesn't really give more detail, but I'm wondering whether the issue is because I'm attempting to run two datasets at once, and these environmental variables is causing issues because they're already 'in use' in the system somehow?  @mikej888 is this a possibility? 

I have a feeling that surely I must have previously had more than one dataset running at once on Eddie previously, but perhaps not since the environmental variables code came in. It is also possible that there's something about the job submission scripts or config files that's to blame, as well as my lack of familiarity running the latest config setups. 

For now, I'm planning to leave the running job to complete (hopefully), then will try submitting another.  This does have implications in terms of knock-on time delays if I can only run them in serial, but I might try running one dataset locally (I successfully ran one sample from L-Sc_2014 earlier today, but it does take a while!) as another test.

**Comment by mikej888** [comment]
@FlicAnderson sorry that's a mystery to me. I could understand it if you were running two Nextflow instances within a single interactive sessions but two batch jobs should surely be using entirely separate nodes.

**Comment by FlicAnderson** [comment]
Somehow the running job on Eddie seemed to have hung up at generateStatsFigs stage (I didn't see any file modifications for the last ~4 hours), so I tried running one of the other jobs instead. Same error.  

Now I can't even get the first job that did start (L-Sc_2014) to run again, despite not changing anything in the job submission script or config files which ran as far as generateStatsFigs before hanging...   Think I'll have to give up with Eddie for tonight...  Not ideal :facepalm: 

I'm not sure how to troubleshoot the error that's happening when I try and run the submitted jobs: 
```
Error occurred during initialization of VM
Could not allocate metaspace: 1073741824 bytes
NOTE: Nextflow is trying to use the Java VM defined by the following environment variables:
 JAVA_CMD: /exports/csce/eddie/biology/groups/wallace_rna/anaconda/envs/riboviz/bin/java
 NXF_OPTS: 
``` 

In 'running things locally' news, it looks like there's deinitely a (?another?) bug...  :beetle: 
For G-Hv_2020, there are 3 samples: FFC1, FFC2, and MNase. 
I ran all samples in 1 go, running on my local laptop: 
```
Completed at: 17-Jun-2021 20:01:45
Duration    : 2h 9m 5s
CPU hours   : 3.9
Succeeded   : 38
```
I get 3 folders in output, and the. pdfs actually look like the ones posted in @acope3 's [comment on the dataset ticket](https://github.com/riboviz/example-datasets/issues/28#issuecomment-733940260).  So far so good. 

I checked the .html outputs: They're all the same.  
They are allocated 1 to each sample folder, and they're named correctly, but once you look at the .html, the Sample Name headings all show "MNase", regardless of which sample it's SUPPOSED to be. 
The plots are all from the MNase sample as well, instead of being from the ones they should be.  

The PDFs show that the correct outputs ARE being created, but it's not doing the right things in `staticHTML` (either the nextflow process, or the AnalysisOutputs side of things, I'm not sure).  I'm not sure if this is the weird *"happens sometimes, I'm not sure why and can't replicate it"* :ghost: :beetle: issue I had from #193 which seemed to have gone?

I didn't notice this issue earlier from the local run because I only ran one L-Sc_2014 sample previously (to test that it would run at all, and then I'd hoped I could move over to running on Eddie :cry: ). 

**The wrong-content .html output issue for G-Hv_2020 AND the weird "Error occurred during initialization of VM" error on Eddie for all 3 datasets have got me stumped for tonight (I have to stop for today now I think), and I think troubleshooting these issues could have rather unfortunate consequences for our 2.1 release schedule** @mikej888 @ewallace @kavousan 

I've started a local machine run of W-Ec_2015, and will check the .html outputs carefully when it completes. Depending on time, I might be able to start L-Sc_2014 (ALL samples this time!) locally before I go to bed.  (For what it's worth, the outputs for the 1-sample run of L-Sc_2014 I got earlier DO match outputs from a run of that dataset I did on the 21st of May or thereabouts, so that's a bit of a positive.)

**Comment by FlicAnderson** [comment]
Having thought a bit more about it and delved into the depths of the 'html output' issue ticket, I think this .html output issue IS the same intermittent bug I discovered previously in developing #193 and [mentioned here](https://github.com/riboviz/riboviz/issues/193#issuecomment-784624146). 

> heavy_minus_sign bug discovered an intermittent bug in my code for staticHTML process (at least I think that's where it's located). Frequently (but frustratingly not always?) the process seems to get the sample_id confused, so both my outputs WT3AT_output_report.html and WTnone_output_report.html contain data for the SAME sample_id.

> Usually (?always?) the sample_id and data in both .html files is that of the second sample processed by staticHTML process. The files are put into their respective sample folders fine, but the content is wrong. Looking at the .command.sh content for each process, the sampleid placed in the rmarkdown::render() command is correct, but inside the markdown html output, print(params$sampleid) shows the wrong value. 


I think that we talked about the possibility of this being a quirk of multi-core shenanigans in my Dell laptop, and I don't think Mike was able to reproduce it at the time.  

So potentially, it might not be a blocker for anyone else.  It could just be the result of a momentary lapse of reason by my processor. I will try and run the same dataset again in the morning and see if the same thing occurs, or if it really is just an intermittent error.  If the latter, then we might still be ok to release?

**Comment by mikej888** [comment]
For
```
Error occurred during initialization of VM
Could not allocate metaspace: 1073741824 bytes
```
Some resources say this relates to `ulimit` which restricts the number of resources for a user. If I run this on my VM I get:
```console
$ ulimit -v
unlimited
```
but I'd guess on EDDIE it might be constrained by whatever the job has asked for. Try adding `ulimit -v` to your job script and see what the value is.

In https://forum.image.sc/t/jvm-error-could-not-allocate-metaspace-solved/12543/5, a user of a Java package in a cluster has exactly the name error with the same number of bytes. A reply commented that:

> the number of bytes refers to the max heap space of the JVM, which by default is 1G or 1073741824 bytes.

(For background, heap space in Java is used for dynamic memory allocation for Java objects and classes when a Java program runs. I've used Java software in the past that complained of running out of memory as too many objects were created and the memory in the heap was consumed. This was solved by setting a Java configuration parameter for it to request a larger heap. The size of the heap is ultimately restricted by the available memory of the host machine.)

The user then comments that:

> Setting the virtual memory in the SGE batch file to this value:
>
> #$ -l h_vmem=6g
>
> fixed it for me.

and then commented that:

> The reported maximum memory usage from the cluster was almost 2G. Apparently you have to at least double the amount of reserved virtual memory, therefore I chose 6G.

So maybe try a larger `h_vmem` value in the job script?

A FAQ, https://rce-docs.hmdc.harvard.edu/faq/why-do-i-get-java-error-could-not-reserve-enough-space-object-heap, also mentions "Could not allocate metaspace" in the context of a research computing environment. They comment that

> To fix the error "Could not allocate metaspace", add the options "-XX:CompressedClassSpaceSize=<size>m" to set the size of the class metadata memory allocation. In testing, 256m has been sufficient.

You should be able to set this via an environment variable in your job script
```
export JAVA_OPTS="-XX:CompressedClassSpaceSize=256m"
export _JAVA_OPTIONS="-XX:CompressedClassSpaceSize=256m"
```
and hopefully that might work.

**Comment by mikej888** [comment]
For the `staticHTML` error I'll run some tests and see if I can replicate it locally. I'll also see if there is a way to find out where the mix up in samples might be arising from (maybe by looking at symbolic links from the `work` directory from `staticHTML` and where these link back to).

**Comment by mikej888** [comment]
For the `staticHTML` error I edited `vignette_config.yaml` to have 4 samples:
```
  WTnone: SRR1042855_s1mi.fastq.gz
  WT3AT: SRR1042864_s1mi.fastq.gz
  NotHere: example_missing_file.fastq.gz # Test case for missing file
=>
  WT1: SRR1042855_s1mi.fastq.gz
  WT2: SRR1042864_s1mi.fastq.gz
  WT3: SRR1042855_s1mi.fastq.gz
  WT4: SRR1042864_s1mi.fastq.gz
```
I ran the workflow but it failed...
```console
$ nextflow run -ansi-log false prep_riboviz.nf -params-file vignette/vignette_config.yaml 
N E X T F L O W  ~  version 20.04.1
Launching `prep_riboviz.nf` [determined_yonath] - revision: f12ad0b907
samples_dir: .
organisms_dir: .
data_dir: .
[a3/d2702c] Submitted process > buildIndicesrRNA (yeast_rRNA)
[35/18a84e] Submitted process > cutAdapters (WT3)
[02/bc240a] Submitted process > buildIndicesORF (YAL_CDS_w_250)
[86/e8dcbb] Submitted process > cutAdapters (WT1)
[3e/96fd7c] Submitted process > cutAdapters (WT4)
[84/ab9772] Submitted process > cutAdapters (WT2)
[15/40958b] Submitted process > createVizParamsConfigFile
[5b/1542bf] Submitted process > hisat2rRNA (WT3)
[2d/c21c2e] Submitted process > hisat2rRNA (WT1)
[a9/d69f5c] Submitted process > hisat2rRNA (WT2)
[48/632c9b] Submitted process > hisat2rRNA (WT4)
[47/6d55e8] Submitted process > hisat2ORF (WT1)
[5c/e80ac6] Submitted process > hisat2ORF (WT3)
[81/2b33a1] Submitted process > trim5pMismatches (WT1)
[f2/c84fb4] Submitted process > trim5pMismatches (WT3)
[8c/3d5fe5] Submitted process > samViewSort (WT1)
[49/efde2f] Submitted process > samViewSort (WT3)
[22/c1ce66] Submitted process > outputBams (WT3)
[68/87455e] Submitted process > outputBams (WT1)
[4b/5d7543] Submitted process > makeBedgraphs (WT3)
[46/212c79] Submitted process > bamToH5 (WT3)
[db/b1b2f8] Submitted process > bamToH5 (WT1)
[d6/ddad35] Submitted process > makeBedgraphs (WT1)
[34/6f4651] Submitted process > hisat2ORF (WT2)
[ca/ade434] Submitted process > hisat2ORF (WT4)
[ea/5bedfc] Submitted process > trim5pMismatches (WT2)
[96/b500a7] Submitted process > trim5pMismatches (WT4)
[e6/0491ac] Submitted process > samViewSort (WT2)
[51/d0f3d8] Submitted process > samViewSort (WT4)
[48/701d7d] Submitted process > outputBams (WT4)
[38/318c2a] Submitted process > outputBams (WT2)
[de/6747be] Submitted process > makeBedgraphs (WT2)
[89/80600c] Submitted process > bamToH5 (WT4)
[29/197d9e] Submitted process > makeBedgraphs (WT4)
[4e/97dd36] Submitted process > bamToH5 (WT2)
[42/a88195] Submitted process > generateStatsFigs (WT1)
[b3/ceafa0] Submitted process > generateStatsFigs (WT3)
[8d/a9cead] Submitted process > generateStatsFigs (WT4)
[d8/18e68e] Submitted process > generateStatsFigs (WT2)
Finished processing sample: WT3
[d6/e26386] Submitted process > renameTpms (WT3)
[bd/6c08e6] Submitted process > staticHTML (WT3)
Finished processing sample: WT1
[2c/6f899c] Submitted process > renameTpms (WT1)
[5f/f3ab67] Submitted process > staticHTML (WT1)
Finished processing sample: WT4
[a0/302859] Submitted process > renameTpms (WT4)
[86/5273d9] Submitted process > staticHTML (WT4)
Finished processing sample: WT2
[00/c0b8dd] Submitted process > renameTpms (WT2)
[52/b5201e] Submitted process > staticHTML (WT2)
[65/76fe79] Submitted process > collateTpms (WT3, WT1, WT4, WT2)
Finished visualising sample: WT3
Finished visualising sample: WT1
[3f/764845] Submitted process > countReads
Finished visualising sample: WT4
Error executing process > 'staticHTML (WT2)'

Caused by:
  Process `staticHTML (WT2)` terminated with an error exit status (1)
...
```
Samples WT1, WT3 and WT4 succeeded but sample WT2 fails during `staticHTML` even though it is the same sample as sample WT4! So I compared the commands run for WT4 and WT2:

```
$ cat work/86/5273d9bbdf9177ef7028201d891aa6/.command.sh
#!/bin/bash -ue
Rscript -e "rmarkdown::render('/home/ubuntu/riboviz/rmarkdown/AnalysisOutputs.Rmd',params = list(verbose='FALSE', yamlfile='$PWD/config.yaml', sampleid='WT4', three_nucleotide_periodicity_data_file = '$PWD/3nt_periodicity.tsv', gene_position_length_counts_5start_file = '$PWD/gene_position_length_counts_5start.tsv', read_length_data_file='$PWD/read_lengths.tsv', pos_sp_rpf_norm_reads_data_file='$PWD/pos_sp_rpf_norm_reads.tsv' , gene_read_frames_filtered_data_file='$PWD/3ntframe_bygene_filtered.tsv', codon_ribodens_gathered_file='$PWD/codon_ribodens_gathered.tsv', sequence_features_file='$PWD/sequence_features.tsv' ), output_format = 'html_document', output_file = '$PWD/WT4_output_report.html')"

$ cat work/52/b5201e9d0b81d96cc7f1d4fdd9a85a/.command.sh 
#!/bin/bash -ue
Rscript -e "rmarkdown::render('/home/ubuntu/riboviz/rmarkdown/AnalysisOutputs.Rmd',params = list(verbose='FALSE', yamlfile='$PWD/config.yaml', sampleid='WT2', three_nucleotide_periodicity_data_file = '$PWD/3nt_periodicity.tsv', gene_position_length_counts_5start_file = '$PWD/gene_position_length_counts_5start.tsv', read_length_data_file='$PWD/read_lengths.tsv', pos_sp_rpf_norm_reads_data_file='$PWD/pos_sp_rpf_norm_reads.tsv' , gene_read_frames_filtered_data_file='$PWD/3ntframe_bygene_filtered.tsv', codon_ribodens_gathered_file='$PWD/codon_ribodens_gathered.tsv', sequence_features_file='$PWD/sequence_features.tsv' ), output_format = 'html_document', output_file = '$PWD/WT2_output_report.html')"
```
and they are the same, except for the sample ID, as expected.

So I compared the logs...

```console
$ cat work/86/5273d9bbdf9177ef7028201d891aa6/.command.log 
...
output file: AnalysisOutputs.knit.md

/usr/bin/pandoc +RTS -K512m -RTS AnalysisOutputs.utf8.md --to html4 --from markdown+autolink_bare_uris+tex_math_single_backslash --output /home/ubuntu/riboviz/work/86/5273d9bbdf9177ef7028201d891aa6/WT4_output_report.html --self-contained --variable bs3=TRUE --standalone --section-divs --table-of-contents --toc-depth 3 --variable toc_float=1 --variable toc_selectors=h1,h2,h3 --variable toc_smooth_scroll=1 --variable toc_print=1 --template /home/ubuntu/R/x86_64-pc-linux-gnu-library/3.6/rmarkdown/rmd/h/default.html --no-highlight --variable highlightjs=1 --variable theme=bootstrap --include-in-header /tmp/RtmpY7pI6Q/rmarkdown-str1c4941f82a22.html --mathjax --variable 'mathjax-url:https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML' 

Output created: WT4_output_report.html
Warning messages:
1: Removed 4 rows containing non-finite values (stat_smooth). 
2: Removed 4 rows containing missing values (geom_point). 
3: Removed 24 rows containing non-finite values (stat_smooth). 
4: Removed 24 rows containing missing values (geom_point). 
```
```console
$ cat work/52/b5201e9d0b81d96cc7f1d4fdd9a85a/.command.log 
...
output file: AnalysisOutputs.knit.md

/usr/bin/pandoc +RTS -K512m -RTS AnalysisOutputs.utf8.md --to html4 --from markdown+autolink_bare_uris+tex_math_single_backslash --output /home/ubuntu/riboviz/work/52/b5201e9d0b81d96cc7f1d4fdd9a85a/WT2_output_report.html --self-contained --variable bs3=TRUE --standalone --section-divs --table-of-contents --toc-depth 3 --variable toc_float=1 --variable toc_selectors=h1,h2,h3 --variable toc_smooth_scroll=1 --variable toc_print=1 --template /home/ubuntu/R/x86_64-pc-linux-gnu-library/3.6/rmarkdown/rmd/h/default.html --no-highlight --variable highlightjs=1 --variable theme=bootstrap --include-in-header /tmp/RtmpbcCrSy/rmarkdown-str1c8c8402c4d.html --mathjax --variable 'mathjax-url:https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML' 
pandoc: AnalysisOutputs.utf8.md: openFile: does not exist (No such file or directory)
Error: pandoc document conversion failed with error 1
In addition: Warning messages:
1: Removed 4 rows containing non-finite values (stat_smooth). 
2: Removed 4 rows containing missing values (geom_point). 
3: Removed 24 rows containing non-finite values (stat_smooth). 
4: Removed 24 rows containing missing values (geom_point). 
Execution halted
```
The failure for WT2 relates to
```
pandoc: AnalysisOutputs.utf8.md: openFile: does not exist (No such file or directory)
```
So the next step is to find out why that file isn't there.

**Comment by mikej888** [comment]
From https://pkgs.rstudio.com/rmarkdown/reference/render.html, `rmarkdown::render` supports two parameters that we don't currently use but which are of interest:

> intermediates_dir	
>
> Intermediate files directory. If a path is specified then intermediate files will be written to that path. If NULL, intermediate files are written to the same directory as the input file.
>
> knit_root_dir	
>
> The working directory in which to knit the document; uses knitr's root.dir knit option. If NULL then the behavior will follow the knitr default, which is to use the parent directory of the document.

Given that `AnalysisOutputs.md` is in `rscripts/` the comments on these parameters imply that any intermediate files will be created in `rscripts/` also. 

A web search provided a useful command, `watch -n 0.1 "ls -ltr"`, which displays the results of `ls -ltr` every 0.1 seconds, so I monitored `rmarkdown/` folder using:

```console
$ watch -n 0.1 "ls -ltr" rmarkdown/
```
I reran the workflow. Two files prefixed by AnalysisOutputs repeatedly appeared and disappeared when rerunning the workflow. Also the workflow completed successfully!

So I'd more strongly suspect now that the reason for the mixed up sample IDs, and, for me, the failed run, is due to `rmarkdown::render` creating intermediate files with the sample name for different samples in the same directory and these are colliding.

Next step, try out those additional `rmarkdown::render` parameters.

**Comment by mikej888** [comment]
I had a look-see on EDDIE as it was a Friday afternoon...

```
Error occurred during initialization of VM
Could not reserve enough space for 16777216KB object heap
```

16777216KB is 16GB.

### 4GB node (`-l h_vmem=4G`)

Request interactive node with 4GB memory for job:
```console
$ qlogin -P bio_wallace_rna_riboviz -l h_vmem=4G
```
Free space on machine:
```console
$ free -h
              total        used        free      shared  buff/cache   available
Mem:           125G         39G         82G        1.6G        3.4G         83G
Swap:           15G        320M         15G
```
Virtual memory resource limit for user:
```console
$ ulimit -v
4194304 
```
4GB as requested in qlogin.

Run `java -version` using `-Xmx` to set maximum Java heap size:
```console
$ java -Xmx1g -version
openjdk version "1.8.0_272"
OpenJDK Runtime Environment (build 1.8.0_272-b10)
OpenJDK 64-Bit Server VM (build 25.272-b10, mixed mode)
$ java -Xmx2g -version
...OK....
$ java -Xmx3g -version
Error occurred during initialization of VM
Could not allocate metaspace: 1073741824 bytes # 1GB
$ java -Xmx4g -version
Error occurred during initialization of VM
Could not reserve enough space for 4194304KB object heap
```
I think the last is expected as it's asking for the entire free memory for the heap.

Try with `-XX:CompressedClassSpaceSize=256m` recommended in https://rce-docs.hmdc.harvard.edu/faq/why-do-i-get-java-error-could-not-reserve-enough-space-object-heap.

A comment on https://bugs.java.com/bugdatabase/view_bug.do?bug_id=8043516 recommends that

> Both XX:MaxHeapSize and XX:CompressedClassSpaceSize have to be specified.

A response to a StackOverflow [question](https://stackoverflow.com/questions/43177466/java-8-xmx-and-xxmaxheapsize-do-not-match) comments that the `-Xmx` option is equivalent to `-XX:MaxHeapSize`.

```console
$ java -Xmx3g -XX:CompressedClassSpaceSize=256m -version
...OK...
```
The 3G heap is now accomodated.
```console
$ java -Xmx4g -XX:CompressedClassSpaceSize=256m -version
Error occurred during initialization of VM
Could not reserve enough space for 4194304KB object heap
```
Set via environment variables:
```console
$ export JAVA_OPTS="-Xmx3g"
$ export _JAVA_OPTIONS="-Xmx3g"
$ java -version
Picked up _JAVA_OPTIONS: -Xmx3g
Error occurred during initialization of VM
Could not allocate metaspace: 1073741824 bytes
$ export JAVA_OPTS="-Xmx3g -XX:CompressedClassSpaceSize=256m"
$ export _JAVA_OPTIONS="-Xmx3g -XX:CompressedClassSpaceSize=256m"
$ java -version
Picked up _JAVA_OPTIONS: -Xmx3g -XX:CompressedClassSpaceSize=256m
...OK...
```

### 16GB node (`-l h_vmem=16G`)

Repeat steps from above but on an interactive node with 16GB memory for job:
```console
$ qlogin -P bio_wallace_rna_riboviz -l h_vmem=16G
```
Free space on machine:
```console
$ free -h
              total        used        free      shared  buff/cache   available
Mem:           125G         39G         82G        1.6G        3.4G         83G
Swap:           15G        320M         15G
```
Virtual memory resource limit for user:
```console
$ ulimit -v
16777216
```
16GB as requested in qlogin.
```console
$ java -Xmx1g -version
openjdk version "1.8.0_272"
OpenJDK Runtime Environment (build 1.8.0_272-b10)
OpenJDK 64-Bit Server VM (build 25.272-b10, mixed mode)
$ java -Xmx2g -version
...OK...
$ java -Xmx3g -version
...OK...
$ java -Xmx4g -version
...OK...
$ java -Xmx8g -version
...OK...
$ java -Xmx12g -version
...OK...
$ java -Xmx14g -version
...OK...
$ java -Xmx15g -version
Error occurred during initialization of VM
Could not allocate metaspace: 1073741824 bytes
```
Try `-XX:CompressedClassSpaceSize=256m`:
```console
$ java -Xmx14g -XX:CompressedClassSpaceSize=256m -version
...OK...
$ java -Xmx15g -XX:CompressedClassSpaceSize=256m -version
...OK...
```
The 15G heap is now accomodated.
```console
$ java -Xmx16g -XX:CompressedClassSpaceSize=256m -version
Error occurred during initialization of VM
Could not reserve enough space for 16777216KB object heap
```

### 4GB node (`-pe interactivemem 4 -l h_vmem=4G`

Use `qlogin` command from `run-on-eddie.md`:

```console
$ qlogin -pe interactivemem 4 -l h_vmem=4G -P bio_wallace_rna_riboviz
$ free -h
              total        used        free      shared  buff/cache   available
Mem:            62G        3.0G         58G        195M        1.3G         58G
Swap:           15G        1.1G         14G
$ ulimit -v
16777216
```
16GB despite 4GB being requested in qlogin. It's because I requested 4 cores.
```console
$ java -Xmx1g -version
openjdk version "1.8.0_272"
OpenJDK Runtime Environment (build 1.8.0_272-b10)
OpenJDK 64-Bit Server VM (build 25.272-b10, mixed mode)
$ java -Xmx2g -version
...OK...
$ java -Xmx3g -version
...OK...
$ java -Xmx4g -version
...OK...
$ java -Xmx8g -version
...OK...
$ java -Xmx12g -version
...OK...
$ java -Xmx14g -version
Error occurred during initialization of VM
Could not allocate metaspace: 1073741824 bytes
$ java -Xmx15g -version
Error occurred during initialization of VM
Unable to allocate 491520KB bitmaps for parallel garbage collection for the requested 15728640KB heap.
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
$ java -Xmx16g -version
Error occurred during initialization of VM
Could not reserve enough space for 16777216KB object heap
```
Try `-XX:CompressedClassSpaceSize=256m`:
```console
$ java -Xmx14g -XX:CompressedClassSpaceSize=256m -version
...OK...
$ java -Xmx15g -XX:CompressedClassSpaceSize=256m -version
Error occurred during initialization of VM
Unable to allocate 491520KB bitmaps for parallel garbage collection for the requested 15728640KB heap.
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
$ java -Xmx16g -XX:CompressedClassSpaceSize=256m -version
Error occurred during initialization of VM
Could not reserve enough space for 16777216KB object heap
```

### Ideas for future exploration

Specify `-Xmx` and `XX:CompressedClassSpaceSize` in `JAVA_OPTS` and `_JAVA_OPTIONS` as `export`ed environment variables in the job script.

Use `-XX:CompressedClassSpaceSize=256m`

Use `-Xmx` (maximum heap size) initially as half the `h_vmem` value.

See what happens!

**Comment by mikej888** [comment]
Raised Investigate Java and Nextflow on EDDIE Could not allocate metaspace issue #375 (and moved my comments there)

**Comment by mikej888** [comment]
@FlicAnderson I've merged the #372 fix into develop and main. It'd be useful if you could complete the tests you'd planned. Once done (and hopefully with no more issues) I think we can then do the 2.1 release.

**Comment by FlicAnderson** [comment]
As discussed at Dev meeting on 30th June, I'm planning to try running again on Eddie for the three datasets I'd aimed for previously, and will have time for this next week (beginning 5th July). This will then allow Mike to build release 2.1

**Comment by 3mma-mack** [comment]
I have been running multiple datasets on eddie when checking upgraded config files. I did not encounter any unusual errors, and any that I did were down to pathing errors. All runs completed successfully once paths had been corrected.

The majority of this was run in the develop branch with commit:
```
commit 106a204bc64525db8de7d18e99d5421be1246394
Merge: 68bbe29 bf06a94
Author: Michael Jackson <michaelj@epcc.ed.ac.uk>
Date:   Mon Jun 21 09:10:32 2021 +0100
```

**Comment by swinterbourne** [comment]
I also ran a series of datasets with updated config files on Eddie. The only error encountered was related to incorrect pathways. All runs were successful and produced the expected outputs. 

The datasets were run from the develop branch with the same commit:  

```
commit 106a204bc64525db8de7d18e99d5421be1246394
Merge: 68bbe29 bf06a94
Author: Michael Jackson <michaelj@epcc.ed.ac.uk>
Date:   Mon Jun 21 09:10:32 2021 +0100
```

**Comment by ewallace** [comment]
I think this is done and can be closed. Many of us have used R2.1, extensively.
